### PR TITLE
Add FlutterDesktopViewGetNativeHandle

### PR DESCRIPTION
**Pull request by Swanseo0**
* Makes all tizen view types be able to get target container handle.

### CODE REVIEW

File: shell/platform/tizen/tizen_window_ecore_wl2.h [hunk 1]
```diff
@@ -34,6 +34,8 @@ class TizenWindowEcoreWl2 : public TizenWindow {
 
   void* GetRenderTarget() override { return ecore_wl2_egl_window_; }
 
+  void* GetRenderTargetContainer() override { return ecore_wl2_window_; }
```
Comment (bbrto21) [inline shell/platform/tizen/tizen_window_ecore_wl2.h:37]: `TizenWindow` has a method for the same purpose. Please consider removing the existing method.
```
virtual void* GetWindowHandle();
```

File: shell/platform/tizen/public/flutter_tizen.h [hunk 2]
```diff
@@ -158,12 +158,14 @@ FLUTTER_EXPORT FlutterDesktopViewRef FlutterDesktopViewCreateFromElmParent(
...
 //
-// Cast the returned void* to Evas_Object*.
+// Cast the returned void*
+// - view elementary    : to Evas_Object*.
+// - window elementary  : to Evas_Object*
+// - window ecore wl2   : to Ecore_Wl2_Window*
 // @warning This API is a work-in-progress and may change.
-FLUTTER_EXPORT void* FlutterDesktopViewGetEvasObject(
-    FlutterDesktopViewRef view);
+FLUTTER_EXPORT void* FlutterDesktopViewGetContainer(FlutterDesktopViewRef view);
```
Comment (bbrto21) [inline shell/platform/tizen/public/flutter_tizen.h:168]: This change affects the embedding of the flutter-tizen. please update the flutter-tizen's engine hash and apply this change later :)

**Review by bbrto21 — COMMENTED** [review]
Thank you for contribution.

**Review by JSUYA — COMMENTED** [review]
Is the name `Container` appropriate?
How about `NativeHandle`? I think we can come up with a better name.

File: shell/platform/tizen/tizen_window_ecore_wl2.h (hunk 1 above)
Comment (Swanseo0) [reply shell/platform/tizen/tizen_window_ecore_wl2.h:37]: Thank you for telling me. I deleted the method that mentioned and modified the part of using this method.

File: shell/platform/tizen/flutter_tizen.cc [hunk 3]
```diff
@@ -74,14 +74,10 @@ void FlutterDesktopEngineShutdown(FlutterDesktopEngineRef engine_ref) {
   delete engine;
 }
 
-void* FlutterDesktopPluginRegistrarGetNativeWindow(
```
Comment (swift-kim) [inline shell/platform/tizen/flutter_tizen.cc:77]: Apparently this API is never used. Why was this added in the first place? cc @bbrto21

Comment (bbrto21) [reply shell/platform/tizen/flutter_tizen.cc:77]: As far as I can remember, these were used in plugins, but over time they didn't use the handle from this at all, and eventually, everywhere that uses this API is gone recently.

Comment (bbrto21) [reply shell/platform/tizen/flutter_tizen.cc:77]: I think we can remove it in this PR or start deprecation step for it.
cc @bwikbs what do you think?

**Comment by Swanseo0** [comment]
> Is the name `Container` appropriate? How about `NativeHandle`? I think we can come up with a better name.

When I read and thought about what you said, `Container` seemed to have a strong content of the Container Object of the view elementary. But `NativeHandle` seems to be more suitable in the sense of the handle of container and window.

File: shell/platform/tizen/flutter_tizen.cc (hunk 3 above)
Comment (Swanseo0) [reply shell/platform/tizen/flutter_tizen.cc:77]: @bbrto21 Will there be a plugin that the API uses? If you don't think it's going to happen, what do you think about deleting the API from this PR?

Comment (bwikbs) [reply shell/platform/tizen/flutter_tizen.cc:77]: According to my memory.. It's prepared for ewk(chromium webview)'s key event handling. 
and.. I'm not sure if it's still necessary or not. :sob:

Comment (bbrto21) [reply shell/platform/tizen/flutter_tizen.cc:77]: If you're not sure, how about leaving it as is?

File: shell/platform/tizen/public/flutter_tizen.h [hunk 4]
```diff
@@ -158,11 +158,14 @@ FLUTTER_EXPORT FlutterDesktopViewRef FlutterDesktopViewCreateFromElmParent(
 // @warning This API is a work-in-progress and may change.
 FLUTTER_EXPORT void FlutterDesktopViewDestroy(FlutterDesktopViewRef view);
 
-// Returns a handle to evas object that the FlutterView is drawn to.
+// Returns a handle that the FlutterView is drawn to.
```
Comment (bbrto21) [inline shell/platform/tizen/public/flutter_tizen.h:161]: How about the expression below?
```suggestion
// Returns a native UI toolkit handle for manipulation in host application.
```

**Review by bbrto21 — APPROVED** [review]
LGTM.  I leave a few more comments below.

File: shell/platform/tizen/flutter_tizen.cc (hunk 3 above)
Comment (JSUYA) [reply shell/platform/tizen/flutter_tizen.cc:77]: FlutterDesktopPluginRegistrarGetNativeHandle and FlutterDesktopViewGetNativeHandle return the same handle.
I think we can remove FlutterDesktopPluginRegistrarGetNativeHandle and use FlutterDesktopViewGetNativeHandle if necessary.

Comment (swift-kim) [reply shell/platform/tizen/flutter_tizen.cc:77]: Consider removing the API if it's unused, or redesign the API so that users can clearly understand the usage.

```cpp
// Bad:
// Returns the native handle of |registrar|?
handle = FlutterDesktopPluginRegistrarGetNativeHandle(registrar);

// Good:
view = FlutterDesktopPluginRegistrarGetView(registrar);
handle = FlutterDesktopViewGetNativeHandle(view);
```

Comment (bbrto21) [reply shell/platform/tizen/flutter_tizen.cc:77]: > Consider removing the API if it's unused, or redesign the API so that users can clearly understand the usage.

I think it would be better to treat this as a different topic. @swift-kim What do you think?

Comment (bbrto21) [reply shell/platform/tizen/flutter_tizen.cc:77]: > Consider removing the API if it's unused, or redesign the API so that users can clearly understand the usage.

@Swanseo0 This conversation still open, please don't forget handle it.

File: shell/platform/tizen/tizen_window_ecore_wl2.h [hunk 5]
```diff
@@ -34,6 +34,8 @@ class TizenWindowEcoreWl2 : public TizenWindow {
 
   void* GetRenderTarget() override { return ecore_wl2_egl_window_; }
 
+  void* GetNativeHandle() override { return ecore_wl2_window_; }
+
   void* GetRenderTargetDisplay() override { return wl2_display_; }
```
Comment (swift-kim) [inline shell/platform/tizen/tizen_window_ecore_wl2.h:39]: Please change the order. `GetRenderTargetDisplay` should be defined as close as possible to `GetRenderTarget`.
```suggestion
  void* GetRenderTargetDisplay() override { return wl2_display_; }

  void* GetNativeHandle() override { return ecore_wl2_window_; }
```

**Review by swift-kim — APPROVED** [review]

File: shell/platform/tizen/tizen_window_ecore_wl2.h (hunk 5 above)
Comment (Swanseo0) [reply shell/platform/tizen/tizen_window_ecore_wl2.h:39]: Applied the modifications that suggested.
tizen_window_element.h also has the same content, so modified it.
Thank you for your proposal.

File: shell/platform/tizen/flutter_tizen.cc (hunk 3 above)
Comment (Swanseo0) [reply shell/platform/tizen/flutter_tizen.cc:85]:
```
view = FlutterDesktopPluginRegistrarGetView(registrar);
handle = FlutterDesktopViewGetNativeHandle(view);
```

Have removed the FlutterdesktopplugInegistrargetnativeHandle function and added the FlutterdesktopplugIngistrargetView function.

Unlike using the existing API, two steps are required, but it seems to be more intuitive for users to use.

**Comment by swift-kim** [comment]
Is the tool side PR ready? Make sure to remove `FlutterDesktopPluginRegistrarGetNativeWindow` from the C# embedding's `Interop.cs`.

**Comment by Swanseo0** [comment]
> Is the tool side PR ready? Make sure to remove `FlutterDesktopPluginRegistrarGetNativeWindow` from the C# embedding's `Interop.cs`.

When this PR enters, a PR that applies modifications to the flutter-tizen repo is prepared.
- Change Engine hash
- Remove FlutterDesktopPluginRegistrarGetNativeWindow
- Add FlutterDesktopPluginRegistrarGetView
- Add FlutterDesktopViewGetNativeHandle

File: shell/platform/tizen/flutter_tizen.cc (hunk 3 above)
Comment (bbrto21) [reply shell/platform/tizen/flutter_tizen.cc:85]: > Consider removing the API if it's unused, or redesign the API so that users can clearly understand the usage.
> 
> ```c++
> // Bad:
> // Returns the native handle of |registrar|?
> handle = FlutterDesktopPluginRegistrarGetNativeHandle(registrar);
> 
> // Good:
> view = FlutterDesktopPluginRegistrarGetView(registrar);
> handle = FlutterDesktopViewGetNativeHandle(view);
> ```

I think `FlutterDesktopViewGetNativeHandle` required by ewk_webview_flutter to show software keyboard.
@Swanseo0 please rebase this PR and make a related PR on flutter-tizen and open it.

Comment (swift-kim) [reply shell/platform/tizen/flutter_tizen.cc:85]: Also change the target branch to `flutter-3.0.5-tizen`.

Comment (Swanseo0) [reply shell/platform/tizen/flutter_tizen.cc:85]: @bbrto21 @swift-kim 
After changing the target branch to flutter-3.0.5-tizen, I rebased this PR.
I haven't opened a PR related to flutter-tizen repo yet, but after this PR is merged, i check the engine hash and apply the hash to the relevant PR.
Is it okay to merge this PR? Or do PR need more modifications? Please tell me your opinion.

Comment (swift-kim) [reply shell/platform/tizen/flutter_tizen.cc:85]: Not okay.

1. Open the tool side PR first.
2. Review the PR.
3. Merge the engine side PR.
4. Update the tool side PR with the new engine hash.
5. Merge the tool side PR.

Comment (bbrto21) [reply shell/platform/tizen/flutter_tizen.cc:85]: 1-1. Open the plugins PR for [this draft](https://github.com/flutter-tizen/plugins/pull/400) in @bwikbs's repository 
2-1. Review the plugins PR.
5-1. Merge the plugins side PR.

[This draft](https://github.com/flutter-tizen/plugins/pull/400) is constantly being tested for new tizen profile and you should not break the build status.
